src/openms/source/ANALYSIS/MAPMATCHING/QTClusterFinder.cpp

```diff
@@ -99,6 +100,132 @@ namespace OpenMS
     // update parameters (dummy)
     setParameters_(1, 1);
 
+    if(use_IDs_)
+    {
+      std::unordered_map<String, std::vector<double>> ided_feat_rts;
+      for (const auto& map : input_maps)
+      {
+        for (const auto& feat : map)
+        {
+          const auto& pepIDs = feat.getPeptideIdentifications();
+          if (!pepIDs.empty())
+          {
+            //TODO sort (lose const) or assume first is best?
+            if (!pepIDs[0].getHits().empty())
+            {
+              const auto it_inserted = ided_feat_rts.emplace(pepIDs[0].getHits()[0].getSequence().toString() + "/" + feat.getCharge(), std::vector<double>{feat.getRT()});
+              if (!it_inserted.second)
+              {
+                //TODO use Running/online statistics (if we can live without median) to save memory
+                it_inserted.first->second.push_back(feat.getRT());
+              }
+            }
+          }
+        }
+      }
+
+      //TODO this does not differentiate between the variety of differences between distinct map pairs. E.g.
+      // differences between map 1 and map 2 might be usually very small (e.g. they are replicates), while
+      // differences between map 1 and map 3 are large, since they are different conditions. But we might lose
+      // robust estimates and use more memory if we split them.
+      std::vector<std::pair<double,std::vector<double>>> medians_diffs;
+      medians_diffs.resize(ided_feat_rts.size());
+      Size c = 0;
+      for (auto& id_rts : ided_feat_rts)
+      {
+        #ifdef DEBUG_QTCLUSTERFINDER
+        std::cout << "Stats for " << id_rts.first << ": ";
+        for (const auto& rt : id_rts.second)
+        {
+          std::cout << rt << ", ";
+        }
+        std::cout << std::endl;
+        #endif
+        auto& rts = id_rts.second;
+        std::sort(rts.begin(),rts.end());
+        medians_diffs[c].first = rts[rts.size()/2];
+        medians_diffs[c].second.reserve(rts.size()-1);
+        Size i = 0;
+        for (const auto& rt : rts)
+        {
+          if (i == rts.size()/2) continue;
+          //TODO would relative diffs solve the RT dependency issue sufficiently? Probably not with non-linear shifts
+          medians_diffs[c].second.push_back(std::fabs(rt-medians_diffs[c].first));
+          i++;
+        }
+        c++;
+      }
+      //TODO we could also bin by equal sized RT bins, or by a fixed RT size
+
+      //TODO could be dependent on nr. of maps (e.g. with 5 maps you get [at least?] 4 diffs per ID already)
+      Size min_nr_diffs_per_bin = 50;
+
+      //TODO check if we can assume sorted
+      std::sort(medians_diffs.begin(), medians_diffs.end());
+
+      Size cnt = 0;
+      vector<double> tmp_diffs;
+      double start_rt = 0.;
+      double iqr_mult = 1.5;
+      double min_tolerance = 20;
```

Comment:
I guess this is ok but could lead to situations where few features in a bin lead to e.g. bins 40, 13, 40 -> 40,20,40
but better (more robust) tolerances would be 40, 40, 40 in this case.
After non-linear alignment the tolerance should be a rather simple linear / non-linear function (maybe a linear fit on the RT tolerance ratios would work?).

Reply:
Hmm good idea, so correct the tolerances again afterwards. Do we have simple code for such a fitting?

Reply:
I think for linear fit we have code. Including more robust versions like RANSAC

Reply:
not worth it as it seems